src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -929,125 +725,316 @@ internal ThreadLocalList(int ownerThreadId)
             /// <param name="result">the removed item</param>
             /// <param name="remove">remove or peek flag</param>
-            internal void Steal(out T result, bool remove)
+            internal bool TrySteal(out T result, bool remove)
             {
-                Node tail = _tail;
-                Debug.Assert(tail != null);
-                if (remove) // Take operation
+                while (true)
                 {
-                    _tail = _tail._prev;
-                    if (_tail != null)
+                    // Fast-path check to see if the list is empty.
+                    if (_headIndex >= _tailIndex)
                     {
-                        _tail._next = null;
+                        result = default(T);
+                        return false;
                     }
-                    else
+
+                    // Anything else requires synchronization.
+                    lock (this)
                     {
-                        _head = null;
+                        // Wait for thread-local operations on the list to subside, if there are
+                        // few enough elements in the list that we could conflict with such operations.
+                        if (Count <= 2)
+                        {
+                            if (_currentOp != (int)ConcurrentBagListOperation.None)
+                            {
+                                var spinner = new SpinWait();
+                                while (_currentOp != (int)ConcurrentBagListOperation.None)
+                                {
+                                    spinner.SpinOnce();
+                                }
+                            }
+
+                            if (Count == 0)
+                            {
+                                result = default(T);
+                                return false;
+                            }
+                        }
+
+                        // Increment head, and ensure read of tail doesn't move before it (full fence).
+                        int head = _headIndex;
+                        Interlocked.Exchange(ref _headIndex, head + 1);
+
+                        // If there's an element to steal, do it.
+                        if (head < _tailIndex)
+                        {
+                            int idx = head & _mask;
+                            result = _array[idx];
+                            if (remove)
+                            {
+                                _stealCount++;
+                                _array[idx] = default(T);
+                            }
+                            return true;
+                        }
+                        else
+                        {
+                            // We contended with the local thread, so restore the head
+                            // and loop around to try again.
+                            _headIndex = head;
```

Comment:
> TrySteal would also see the list as empty and fail fast

TrySteal would need to take the lock, which would force it to synchronize with the other steal, and it would end up seeing consistent state.